docs/MV3.md

```diff
@@ -0,0 +1,114 @@
+      'xmlhttprequest'
+    ]
+  }
+}
+```
+
+The API only allows for 5000 such rules to exist at the time of writing this, so that value can be exhausted pretty quickly. So instead of adding one rule per redirect, we can dynamically generate a single regex pattern to handle multiple redirections.
+
+For example
+
+```
+https://ipfs.io/ipns/en.wikipedia-on-ipfs.org -> http://localhost:8080/ipns/en.wikipedia-on-ipfs.org
+```
+
+can be represented as
+
+```
+https://ipfs.io/ipns/(.*) -> http://localhost:8080/ipns/$1
+```
+
+That's just an example, but in reality it looks more like (which covers edge cases like URL endings):
+
+```js
+{
+  "action": {
+    "redirect": {
+      "regexSubstitution": "http://localhost:8080\\1"
+    },
+    "type": "redirect"
+  },
+  "condition": {
+    "excludedInitiatorDomains": [],
+    "regexFilter": "^https?\\:\\/\\/ipfs\\.io((?:[^\\.]|$).*)$",
+    "resourceTypes": [
+      "csp_report",
+      "font",
+      "image",
+      "main_frame",
+      "media",
+      "object",
+      "other",
+      "ping",
+      "script",
+      "stylesheet",
+      "sub_frame",
+      "webbundle",
+      "websocket",
+      "webtransport",
+      "xmlhttprequest"
+    ]
+  },
+  "id": 1234,
+  "priority": 1
+}
+```
+
+This is a single rule that covers all the redirects from `ipfs.io` to `localhost:8080`. This means that we can save a lot of rules and gain more coverage of serviceable URLs. This is the approach we have taken in MV3.
+
+## Gotchas
+
+Since the process is asynchronous, there are a few things that we need to keep in mind:
+
+- There will always be scenarios, where the redirection to public URL happens before companion determines if the URL is serviceable (consider ipns/fqdn resolutions.) To tackle this a refresh mechanism has been put in place that refreshes the page after a few seconds if the URL is serviceable. This is not ideal, but it is the best we can do for now.
+- Similarly there are scenarios around page recovery, where the page starts redirection, but the kubo node goes offline, in which case companion needs to remove offending rule and replace it with the recovery URL.
+- If the kubo rpc api url changes for some reason, companion needs to update the rules to reflect the new URL. We can preemptively generate new rules based on the new url, but the new node might not be servicing the old content, so we need to be careful about that. Hence, we remove all such rules as offending and begin the process of generating new rules dynamically.
+- Metrics collection becomes harder, as we can no longer reliably determine if the requested resource was serviced by declarative rule introduced by companion. We can only determine if the request was serviceable and we had a rule. We can probably introduce a content script to every page and redirect with a hash value in the URL. That content script can then invoke a companion API to notify what rule id was used to redirect the request. This is a future improvement.
```

Comment:
This doesn't seem like a high priority to get tighter metrics here, but if there is a backlog item for it, lets link to it.